src/rspq/rspq.c

```diff
@@ -0,0 +1,1235 @@
+    if (rspq_block) {
+        // Allocate next chunk (double the size of the current one).
+        // We use doubling here to reduce overheads for large blocks
+        // and at the same time start small.
+        if (rspq_block_size < RSPQ_BLOCK_MAX_SIZE) rspq_block_size *= 2;
+
+        // Allocate a new chunk of the block and switch to it.
+        uint32_t *rspq2 = malloc_uncached(rspq_block_size*sizeof(uint32_t));
+        volatile uint32_t *prev = rspq_switch_buffer(rspq2, rspq_block_size, true);
+
+        // Terminate the previous chunk with a JUMP op to the new chunk.
+        rspq_append1(prev, RSPQ_CMD_JUMP, PhysicalAddr(rspq2));
+        return;
+    }
+
+    // Wait until the previous buffer is executed by the RSP.
+    // We cannot write to it if it's still being executed.
+    // FIXME: this should probably transition to a sync-point,
+    // so that the kernel can switch away while waiting. Even
+    // if the overhead of an interrupt is obviously higher.
+    MEMORY_BARRIER();
+    if (!(*SP_STATUS & rspq_ctx->sp_status_bufdone)) {
+        rspq_flush_internal();
+        RSP_WAIT_LOOP(200) {
+            if (*SP_STATUS & rspq_ctx->sp_status_bufdone)
+                break;
+        }
+    }
+    MEMORY_BARRIER();
+    *SP_STATUS = rspq_ctx->sp_wstatus_clear_bufdone;
+    MEMORY_BARRIER();
+
+    // Switch current buffer
+    rspq_ctx->buf_idx = 1-rspq_ctx->buf_idx;
+    uint32_t *new = rspq_ctx->buffers[rspq_ctx->buf_idx];
+    volatile uint32_t *prev = rspq_switch_buffer(new, rspq_ctx->buf_size, true);
+
+    // Terminate the previous buffer with an op to set SIG_BUFDONE
+    // (to notify when the RSP finishes the buffer), plus a jump to
+    // the new buffer.
+    rspq_append1(prev, RSPQ_CMD_WRITE_STATUS, rspq_ctx->sp_wstatus_set_bufdone);
+    rspq_append1(prev, RSPQ_CMD_JUMP, PhysicalAddr(new));
+    assert(prev+1 < (uint32_t*)(rspq_ctx->buffers[1-rspq_ctx->buf_idx]) + rspq_ctx->buf_size);
+    rspq_flush_internal();
+}
+
+__attribute__((noinline))
+static void rspq_flush_internal(void)
+{
+    // Tell the RSP to wake up because there is more data pending.
+    MEMORY_BARRIER();
+    *SP_STATUS = SP_WSTATUS_SET_SIG_MORE | SP_WSTATUS_CLEAR_HALT | SP_WSTATUS_CLEAR_BROKE;
+    MEMORY_BARRIER();
+
+    // Most of the times, the above is enough. But there is a small and very rare
+    // race condition that can happen: if the above status change happens
+    // exactly in the few instructions between RSP checking for the status
+    // register ("mfc0 t0, COP0_SP_STATUS") RSP halting itself("break"),
+    // the call to rspq_flush might have no effect (see command_wait_new_input in
+    // rsp_queue.S).
+    // In general this is not a big problem even if it happens, as the RSP
+    // would wake up at the next flush anyway, but we guarantee that rspq_flush
+    // does actually make the RSP finish the current buffer. To keep this
+    // invariant, we wait 10 cycles and then issue the command again. This
```

Comment:
I think this explains my concerns there :)

Reply:
Can you elaborate? In case you are concerned about the fact that this "hack" is even necessary, it's pretty much the only way to prevent a race condition (see my other comment in `RSPQCmd_WaitNewInput`). At least to my knowledge.

Reply:
I was referring to my comment at https://github.com/DragonMinded/libdragon/pull/253/files/2feb4242db3097518ca09a28b5c373ff73597083#diff-8f33e64dc050f62d7f9988d1952992863c56dca0a901e9bddd099c76825e816bR237 When reading the PR, this actually answered my question there, as you have also mentioned `RSPQCmd_WaitNewInput`